File: backend/src/routes/web.js
import { Router } from 'express';
import { requireAuth } from '../middleware/auth.js';

export const webRouter = Router();
webRouter.use(requireAuth);

// Search DuckDuckGo (no API key needed)
webRouter.get('/web/search', async (req, res) => {
  try {
    const q = req.query.q;
    if (!q) return res.status(400).json({ error: 'Query required' });

    // Use DuckDuckGo lite/HTML endpoint
    const url = `https://lite.duckduckgo.com/lite/?q=${encodeURIComponent(q)}`;
    const response = await fetch(url, {
      headers: { 'User-Agent': 'Mozilla/5.0 (compatible; HAgent/1.0)' },
    });
    const html = await response.text();

    // Parse results from HTML
    const results = [];
    const linkRegex = /<a[^>]*href="([^"]*)"[^>]*class="result-link"[^>]*>(.*?)<\/a>/gi;
    const snippetRegex = /<td[^>]*class="result-snippet"[^>]*>(.*?)<\/td>/gi;
    const urlRegex = /<span[^>]*class="result-url"[^>]*>(.*?)<\/span>/gi;

    // Alternative parsing - simpler regex
    const rows = html.split('<tr>');
    for (let i = 0; i < rows.length; i++) {
      const row = rows[i];
      const titleMatch = row.match(/class="result-link"[^>]*>\s*(.*?)\s*<\/a>/i);
      const snippetMatch = row.match(/class="result-snippet"[^>]*>\s*(.*?)\s*<\/td>/i);
      const urlMatch = row.match(/class="result-url"[^>]*>\s*(.*?)\s*<\/span>/i);

      if (titleMatch && urlMatch) {
        results.push({
          title: titleMatch[1].replace(/<[^>]*>/g, '').trim(),
          url: urlMatch[1].replace(/<[^>]*>/g, '').trim(),
          snippet: snippetMatch ? snippetMatch[1].replace(/<[^>]*>/g, '').trim() : '',
        });
      }
    }

    res.json({ query: q, results: results.slice(0, 10) });
  } catch (err) {
    console.error('Search error:', err);
    res.status(500).json({ error: err.message });
  }
});

// Fetch a URL and extract text content
webRouter.post('/web/fetch', async (req, res) => {
  try {
    const { url } = req.body;
    if (!url) return res.status(400).json({ error: 'URL required' });

    const response = await fetch(url, {
      headers: { 'User-Agent': 'Mozilla/5.0 (compatible; HAgent/1.0)' },
      signal: AbortSignal.timeout(10000),
    });

    const html = await response.text();

    // Extract text content (strip HTML tags)
    const text = html
      .replace(/<script[^>]*>[\s\S]*?<\/script>/gi, ' ')
      .replace(/<style[^>]*>[\s\S]*?<\/style>/gi, ' ')
      .replace(/<[^>]*>/g, ' ')
      .replace(/&[^;]+;/g, ' ')
      .replace(/\s+/g, ' ')
      .trim()
      .slice(0, 5000);

    res.json({ url, text });
  } catch (err) {
    console.error('Fetch error:', err);
    res.status(500).json({ error: err.message });
  }
});

File: backend/api/routers/web.py
"""Web router — search and fetch, delegating to tools/web_tools.py."""

import re
from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel

router = APIRouter(prefix="/api/web", tags=["web"])


def _import_web_tools():
    import sys
    # web_tools.py has heavy imports; lazy-load on first call
    if "_web_search_fn" not in dir():
        from tools.web_tools import web_search_tool
        return web_search_tool
    return None


@router.get("/search")
async def search(request: Request, q: str = ""):
    if not q:
        raise HTTPException(status_code=400, detail="Query required")
    try:
        from tools.web_tools import web_search_tool
        results = web_search_tool(q, limit=10)
        # Parse the text output into structured results
        lines = results.split("\n")
        parsed = []
        for line in lines:
            if line.startswith("  ") or not line.strip():
                continue
            parsed.append({"title": line.strip(), "url": "", "snippet": ""})
        return {"query": q, "results": parsed[:10]}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


class FetchBody(BaseModel):
    url: str


@router.post("/fetch")
async def fetch_url(body: FetchBody):
    if not body.url:
        raise HTTPException(status_code=400, detail="URL required")
    try:
        import aiohttp
        async with aiohttp.ClientSession(headers={"User-Agent": "Mozilla/5.0 (compatible; HAgent/1.0)"}) as s:
            async with s.get(body.url, timeout=aiohttp.ClientTimeout(total=10)) as r:
                html = await r.text()
        text = re.sub(r'<script[^>]*>.*?</script>', '', html, flags=re.DOTALL)
        text = re.sub(r'<style[^>]*>.*?</style>', '', text, flags=re.DOTALL)
        text = re.sub(r'<[^>]*>', ' ', text)
        text = re.sub(r'&[^;]+;', ' ', text)
        text = re.sub(r'\s+', ' ', text).strip()[:5000]
        return {"url": body.url, "text": text}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
